### ISSUE TITLE
BUG in cuda:1

### ISSUE THREAD

**Issue by bensonbs** [ultralytics/ultralytics#4556]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Bug

```
RuntimeError: Expected all tensors to be on the same device, but found at least two devices, cuda:0 and cuda:1! (when checking argument for argument weight in method wrapper_CUDA__cudnn_convolution)
```

### Environment

OS: Windows 11
torch: '2.0.1+cu117'
ultralytics: '8.0.160'

### Additional

```
model = YOLO(f'./models/best.pt')

transform = transforms.Compose([
    Resize((640, 640)),
    transforms.ToTensor(),
])

```

move model to `cuda:1`, but after predict model move to `cuda:0`
```
img_tensor = transform(Image.fromarray(image))
img_cuda = img_tensor.unsqueeze(0).to('cuda:1')

model.to('cuda:1')
>>> device(type='cuda', index=1)

model(img_cuda)

model.device
>>>device(type='cuda', index=0) <-- index get "0"
```

let model in to `cuda:1` again
```
model.to('cuda:1')
model.device
>>> device(type='cuda', index=1)
```

then predict again get the error
```
img_cuda.device
>>>device(type='cuda', index=1)

model(img_cuda)
>>>RuntimeError: Expected all tensors to be on the same device, but found at least two devices, cuda:0 and cuda:1! (when checking argument for argument weight in method wrapper_CUDA__cudnn_convolution)

```

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@bensonbs hello, 

Thank you for providing a concise bug report with a minimal reproducible example. We appreciate your help in understanding the issue at hand!

The error message you're seeing implies that the device is being switched from `cuda:1` to `cuda:0` during the prediction process.

In your script, you are rightfully setting the model and input image tensor to the same device, `cuda:1`. It looks like the device for our model isn't persisting and switches back to `cuda:0` after prediction. This mismatch in devices between the model and the input tensor gives rise to the error you're seeing when you try to predict again.

Of course, proper behavior for YOLOv8 should be to persist the device assignment for the model until explicitly changed by the user. We will need to investigate why the model's device is switching back to `cuda:0` after prediction. 

I understand the difficulty this issue can cause in your work, and I appreciate your patience as we look into resolving this. We will keep you updated regarding any progress or findings regarding this issue. 

And while you've unchecked the option to submit a PR, we’d always appreciate and welcome any collaborative efforts from our community if you feel inclined to get involved.

**Comment by nouranali** [comment]
I can start investigating why this happens, do you have a suggestion from where can I start? @glenn-jocher

**Comment by bensonbs** [comment]
https://github.com/ultralytics/ultralytics/issues/4556#issuecomment-1722874125

I think it is a problem with pytorch. I can use version 1.13.1 of pytorch without triggering this bug.

**Comment by glenn-jocher** [comment]
@nouranali hi, thanks for sharing your observations. It seems like you've found a workaround by switching to PyTorch version 1.13.1. It's possible that the issue is linked to changes in later versions of PyTorch regarding how tensors and models are handled across different devices.

Your insight is valuable and we'll take this into account as we continue to investigate the problem. If you discover more or have any other updates, feel free to share them. Thanks for helping improve YOLOv8!

**Comment by Mengb001** [comment]
I encountered the same error, my torch version is 1.12 and I changed the code for class modelEMA's def update。

    def update(self, model):
        """Update EMA parameters."""
        if self.enabled:
            self.updates += 1
            d = self.decay(self.updates)

            msd = de_parallel(model).state_dict()  # model state_dict
            for k, v in self.ema.state_dict().items():
                if v.dtype.is_floating_point:  # true for FP16 and FP32
                    v *= d
                    msd[k] = msd[k].to(v.device)                       #######fix bug 
                    v += (1 - d) * msd[k].detach()

**Comment by Mengb001** [comment]
Should I use Torch version 1.13?

**Comment by glenn-jocher** [comment]
@Mengb001 it's a good idea to try using PyTorch version 1.13.1. A user reported the same bug and found that switching to PyTorch version 1.13.1 wasn't triggering this issue. However, it's also important to note that downgrading the PyTorch version is a workaround, not a solution. Meanwhile, our team will keep investigating this issue. Please keep posted for updates on a permanent solution.

**Comment by Mengb001** [comment]
> @Mengb001 it's a good idea to try using PyTorch version 1.13.1. A user reported the same bug and found that switching to PyTorch version 1.13.1 wasn't triggering this issue. However, it's also important to note that downgrading the PyTorch version is a workaround, not a solution. Meanwhile, our team will keep investigating this issue. Please keep posted for updates on a permanent solution.

Thank you for your elegant work. Also, how do I get V8 latency and FPS?

**Comment by glenn-jocher** [comment]
Hi @Mengb001, thank you for your kind words.

Latency is typically measured by the time taken for an image or batch of images to pass through the network (i.e., forward pass time). In your code, you can use Python's built-in time module to record the time just before and just after the forward pass, and then compute the difference.

Frames per second (FPS) is usually computed as the reciprocal of the time taken for a forward pass of each frame.

For instance, if it takes 0.05 seconds for an image to pass through the network, your latency would be 0.05 seconds, and your FPS would be the reciprocal of that, which is 20 FPS.

Regarding your previous question, we're glad that PyTorch 1.13.1 is working for you as a temporary workaround. Just to clarify, downgrading the PyTorch version is simply a way to sidestep the issue for now, rather than a formal solution. Rest assured, our team is continuing to investigate this issue and we'll certainly keep you updated on any further discoveries or potential fixes.

**Comment by Mengb001** [comment]
Hi @glenn-jocher  Thank you for your answer！
I have another question：
How do I get metrics for my model in COCO format on the V8？like ap(s)  ap(m)  ap(L)

**Comment by glenn-jocher** [comment]
@Mengb001 hello, 

To get detailed evaluation metrics like AP(small), AP(medium), and AP(large) in COCO format on YOLOv8, you need to use the COCO API, which provides tools for dataset parsing and evaluation. 

First, you need to format your prediction results to match the COCO format. This could be a .json file with bounding box coordinates, categories, and confidence scores - structured as per the COCO specifications. 

Then, you load this file with your predictions and the ground truth annotations into the API. Functions provided by the API like COCOeval can be used to compute the metrics that you mentioned. 

The results usually give an overview of the performance of the model for different object sizes segmented as small, medium, and large. 

Just make sure to handle the class IDs correctly if you're not using the standard COCO dataset but a custom one. It's important that the predicted class IDs and the ground truth class IDs match for the evaluation to be correct. 

I hope this helps! If you have any other questions, feel free to ask.

**Comment by Mengb001** [comment]
@glenn-jocher Thank you so much for your wonderful work.
In addition，I have some questions about the configuration file for RTDETR.
Below is Baidu's profile, with a focus on layer num：
'stage_config': {
                # in_channels, mid_channels, out_channels, num_blocks, downsample, light_block, kernel_size, layer_num
                "stage1": [48, 48, 128, 1, False, False, 3, 6],
                "stage2": [128, 96, 512, 1, True, False, 3, 6],
                "stage3": [512, 192, 1024, 3, True, True, 5, 6],
                "stage4": [1024, 384, 2048, 1, True, True, 5, 6],
            }
Here is our configuration file, focusing on layer num：
 # [from, repeats, module, args]
  - [-1, 1, HGStem, [32, 48]]  # 0-P2/4
  - [-1, 6, HGBlock, [48, 128, 3，1]]  # stage 1

  - [-1, 1, DWConv, [128, 3, 2, 1, False]]  # 2-P3/8
  - [-1, 6, HGBlock, [96, 512, 3，1]]   # stage 2

  - [-1, 1, DWConv, [512, 3, 2, 1, False]]  # 4-P3/16
  - [-1, 6, HGBlock, [192, 1024, 5, 1，True, False]]  # cm, c2, k, light, shortcut
  - [-1, 6, HGBlock, [192, 1024, 5, 1， True, True]]
  - [-1, 6, HGBlock, [192, 1024, 5, 1，True, True]]  # stage 3

  - [-1, 1, DWConv, [1024, 3, 2, 1, False]]  # 8-P4/32
  - [-1, 6, HGBlock, [384, 2048, 5, 1，True, False]]  # stage 4
  Baidu's HGblock has a layer num of 6, our HGblock has a layer num of 1 by default, but repeats is 6. Will there be problems with our profile?

**Comment by glenn-jocher** [comment]
@Mengb001 hello, thank you for pointing out the differences in configuration for the 'layer_num' parameter in the Baidu's RTDETR configuration and the YOLOv8 configuration.

In YOLOv8's configuration, the 'layer_num' value is controlled by the 'repeats' argument and is set to 6, in line with the Baidu's configuration where 'layer_num' is also set to 6. 

This difference in 'repeats' versus 'layer_num' naming is a distinction between YOLOv8 and Baidu's system. The key point is the number of repeats or layers being used. In both the configurations, the structures are repeated the same number of times (6 times), so there should not be any inconsistencies or problems due to this.

In summary, both configurations essentially achieve the same end result, so the YOLOv8 configuration you provided should work as expected.

**Comment by Mengb001** [comment]
@glenn-jocher I'm still a little confused. Maybe I didn't understand the point。In my opinion，Baidu's structure looks like this：
class HGBlock(nn.Module):
    """
    HG_Block of PPHGNetV2 with 2 convolutions and LightConv.

    https://github.com/PaddlePaddle/PaddleDetection/blob/develop/ppdet/modeling/backbones/hgnet_v2.py
    """

    def __init__(self, c1, cm, c2, k=3, n=6, lightconv=False, shortcut=False, act=nn.ReLU()):
        """Initializes a CSP Bottleneck with 1 convolution using specified input and output channels."""
        super().__init__()
        block = LightConv if lightconv else Conv
        self.m = nn.ModuleList(block(c1 if i == 0 else cm, cm, k=k, act=act) for i in range(n))
        self.sc = Conv(c1 + n * cm, c2 // 2, 1, 1, act=act)  # squeeze conv
        self.ec = Conv(c2 // 2, c2, 1, 1, act=act)  # excitation conv
        self.add = shortcut and c1 == c2

    def forward(self, x):
        """Forward pass of a PPHGNetV2 backbone layer."""
        y = [x]
        y.extend(m(y[-1]) for m in self.m)
        y = self.ec(self.sc(torch.cat(y, 1)))
        return y + x if self.add else y


HGblock                 （conv→conv→conv→conv→conv→conv）
                                      ⬇    ⬇       ⬇    ⬇        ⬇    ⬇
                                                              cat
                                                              ⬇
                                                              SCconv
                                                                ⬇
                                                               ECconv
The structure of V8 looks like this：
                           （conv→SCconv→ECconv）→（conv→SCconv→ECconv）→（conv→SCconv→ECconv）→（conv→SCconv→ECconv）→（conv→SCconv→ECconv）→（conv→SCconv→ECconv）

Excuse me, what did I understand wrong? But I understand that your team's RTDETR params are correct.
Finally, it's been a pleasure communicating with you

**Comment by glenn-jocher** [comment]
Hi @Mengb001, I appreciate your detailed consideration of the configurations and architectures. The differences you highlighted between the Baidu stack of convolutions and YOLOv8's sequential blocks illustrate a subtle yet essential aspect of network design.

In the Baidu's `HGBlock` implementation, as described, multiple convolutional layers (represented by `self.m`) are applied sequentially, where the output of each convolution is stored and then concatenated before being passed through a squeeze (`self.sc`) and an excitation (`self.ec`) convolution. This suggests a DenseNet-style connectivity where each layer receives inputs from all preceding layers.

For YOLOv8, the successive repeats of `HGBlock` suggest that each block conducts its own processing (`conv→SCconv→ECconv`), and then that output serves as input to the next block in a sequential manner rather than the concatenation approach. 

What's important is that regardless of the internal structure, both the Baidu and YOLOv8 configuration are repeat units of their fundamental block, aiming to deepen the network and expand the receptive field while maintaining a balance between efficiency and complexity. Each 'repeat' in YOLOv8's configuration file signifies an entire HGBlock unit with its internal layers and operations.

The depth (number of layers) in both cases (6 for Baidu's stack and 6 repeats for YOLOv8) is the number of these internal units used in the model, and if their total depth is equivalent, they will potentially yield similar representational power. However, the internal features and operations might yield different characteristics due to their architectural differences.

The Ultralytics team has tailored the hierarchical blocks in the YOLOv8 to be efficient and effective for the given tasks, building upon prior art while optimizing for certain conditions. The correctness of YOLOv8's `HGBlock` configuration depends upon the integration within the whole model and alignment with the intended design philosophy.

It's been great discussing these technical details with you. Your questions demonstrate a deep engagement with the domain, and it's a pleasure to assist you with any clarifications. If you have further inquiries or dive deeper into other aspects of YOLOv8 or related works, feel free to reach out.

**Comment by sarpx** [comment]
@glenn-jocher @bensonbs 

I am experiencing this problem in different ways.

1. You have a picture and you want to look at the result in two different models. It definitely doesn't work. And it gives the error 'Image' object has no attribute 'shape'.

2. You are providing training, the training has been interrupted. Then you increased the milk level and the unfinished education cannot reach the cuda. You need to train from scratch.

When I researched, I found something like this. I think the solution might be to include it in yolo.

https://discuss.pytorch.org/t/attributeerror-image-object-has-no-attribute-shape/149175/2

**Comment by glenn-jocher** [comment]
@sarpx thank you for reporting this issue and providing detailed information. It seems like there might be an unexpected behavior where the model's device is being reset to `cuda:0` after inference. 

As a quick workaround, you can try explicitly setting the device before each prediction:

```python
model.to('cuda:1')
results = model(img_cuda)
```

However, this is not a permanent solution. We will investigate this issue further to understand why the device assignment is changing after prediction. In the meantime, if you have any additional information or steps that could help us replicate the issue, please share them.

If you find a solution or would like to contribute to fixing this bug, we welcome pull requests. Otherwise, we will update you once we have a fix. Thank you for your patience! 🛠️

**Comment by sarpx** [comment]
I understood what the problem was. You cannot use models trained with old versions together with another model trained with the new version. This probably has something to do with CUDA.
If you are interested, there is a comment below where I explain the details in another question.

https://github.com/ultralytics/ultralytics/issues/8073#issuecomment-1932563094

**Comment by Y-T-G** [comment]
Seems to longer be an issue in latest Ultralytics
